core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java

```diff
@@ -73,9 +73,10 @@ public void testMasterFailoverDuringIndexingWithMappingChanges() throws Throwabl
                 .put("discovery.zen.join_timeout", "10s")  // still long to induce failures but to long so test won't time out
                 .put(DiscoverySettings.PUBLISH_TIMEOUT_SETTING.getKey(), "1s") // <-- for hitting simulated network failures quickly
                 .put(ElectMasterService.DISCOVERY_ZEN_MINIMUM_MASTER_NODES_SETTING.getKey(), 2)
+                .put(DiscoverySettings.INITIAL_STATE_TIMEOUT_SETTING.getKey(), "0")  // <-- we wait for cluster formation at the end
```

Comment:
> That method does very little and is mostly about scheduling some stuff on the thread pool and starting the joining process (all of which is done in separate threads anyhow).

Agreed and I see that side of the argument as well. We discussed this and the majority feels that limiting the async to the start method is the right trade off (in order to not have custom settings). I'll adapt the PR.